src/config.cpp

```diff
@@ -478,5 +479,127 @@ const std::string& CircuitConfig::getExpandedJSON() const {
+        parseMandatory(*runIt, "dt", "run", result.dt);
+        return result;
+    }
+
+    SimulationConfig::Output parseOutput() const {
+        SimulationConfig::Output result{};
+        result.outputDir = "output";
+        result.spikesFile = "out.h5";
+
+        const auto outputIt = _json.find("output");
+        if (outputIt != _json.end()) {
+            parseOptional(*outputIt, "output_dir", result.outputDir);
+            parseOptional(*outputIt, "spikes_file", result.spikesFile);
+        }
+
+        result.outputDir = toAbsolute(_basePath, result.outputDir);
+
+        return result;
+    }
+
+    std::unordered_map<std::string, SimulationConfig::Report> parseReports() const {
+        std::unordered_map<std::string, SimulationConfig::Report> result;
+
+        const auto reportsIt = _json.find("reports");
+        if (reportsIt == _json.end())
+            return result;
+
+        for (auto it = reportsIt->begin(); it != reportsIt->end(); ++it) {
+            auto& report = result[it.key()];
+            auto& valueIt = it.value();
+            const auto debugStr = fmt::format("report {}", it.key());
+            parseMandatory(valueIt, "cells", debugStr, report.cells);
+            parseMandatory(valueIt, "type", debugStr, report.type);
+            parseMandatory(valueIt, "dt", debugStr, report.dt);
+            parseMandatory(valueIt, "start_time", debugStr, report.startTime);
+            parseMandatory(valueIt, "end_time", debugStr, report.endTime);
+            parseOptional(valueIt, "file_name", report.fileName);
+            if (report.fileName.empty())
+                report.fileName = it.key() + "_SONATA.h5";
+            else if (report.fileName.find(".h5") == std::string::npos)
+                report.fileName += ".h5";
+            report.fileName = toAbsolute(_basePath, report.fileName);
+        }
+
+        return result;
+    }
+
+  private:
+    const fs::path _basePath;
+    const nlohmann::json _json;
+};
+
+SimulationConfig::SimulationConfig(const std::string& content, const std::string& basePath)
+    : _jsonContent(content)
+    , _basePath(fs::absolute(basePath).lexically_normal().string()) {
+    const Parser parser(content, basePath);
+    _run = parser.parseRun();
+    _output = parser.parseOutput();
+    _reports = parser.parseReports();
+}
+
+SimulationConfig SimulationConfig::fromFile(const std::string& path) {
+    return SimulationConfig(readFile(path), fs::path(path).parent_path());
+}
+
+const std::string& SimulationConfig::getBasePath() const noexcept {
+    return _basePath;
+}
+
+const std::string& SimulationConfig::getJSON() const noexcept {
+    return _jsonContent;
+}
+
+const SimulationConfig::Run& SimulationConfig::getRun() const noexcept {
+    return _run;
+}
+
+const SimulationConfig::Output& SimulationConfig::getOutput() const noexcept {
+    return _output;
+}
+
+const SimulationConfig::Report& SimulationConfig::getReport(const std::string& name) const {
```

Comment:
It's probably not worth doing anything about this now...but once C++17/20 are available then using `std::string_view` and transparent comparators might be nicer.

Reply:
I agree, although that would be more of a library-wide change